compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELogicalObjFifoSplittingUtils.cpp

```diff
@@ -427,4 +476,386 @@ LogicalResult splitLogicalObjectFifos(
+  }
+  return false;
+}
+
+static LogicalResult checkIfSameDimensionalityAccessPatterns(
+    AMDAIE::DmaCpyNdOp &l3ToL2DmaOpA, AMDAIE::DmaCpyNdOp &l3ToL2DmaOpB) {
+  SmallVector<OpFoldResult> sourceOffsetsA =
+      l3ToL2DmaOpA.getSourceMixedOffsets();
+  SmallVector<OpFoldResult> sourceSizesA = l3ToL2DmaOpA.getSourceMixedSizes();
+  SmallVector<OpFoldResult> sourceStridesA =
+      l3ToL2DmaOpA.getSourceMixedStrides();
+  SmallVector<OpFoldResult> sourceOffsetsB =
+      l3ToL2DmaOpB.getSourceMixedOffsets();
+  SmallVector<OpFoldResult> sourceSizesB = l3ToL2DmaOpB.getSourceMixedSizes();
+  SmallVector<OpFoldResult> sourceStridesB =
+      l3ToL2DmaOpB.getSourceMixedStrides();
+  if (sourceOffsetsA.size() != sourceOffsetsB.size() ||
+      sourceSizesA.size() != sourceSizesB.size() ||
+      sourceStridesA.size() != sourceStridesB.size() ||
+      sourceOffsetsA.size() != sourceSizesA.size() ||
+      sourceOffsetsA.size() != sourceStridesB.size()) {
+    return failure();
+  }
+  return success();
+}
+
+/// Given a vector of L2->L1 Dma Ops, combine the corresponding L3->L2 Dma Ops
+/// and reuse the L2/L1 buffers.
+/// TODO(avarma): Assign combined tiles while forming L2/L1 buffers which we'll
+/// reuse.
+LogicalResult combineLogicalObjectFifos(
+    IRRewriter &rewriter, SmallVector<AMDAIE::DmaCpyNdOp> &l2ToL1DmaOps,
+    MLIRContext *context) {
+  if (l2ToL1DmaOps.size() == 0) return success();
+
+  // Fetch the L3 -> L2 Dma Op corresponding to the first L2 buffer as target.
+  SmallVector<AMDAIE::DmaCpyNdOp> l3ToL2DmaOps;
+  FailureOr<AMDAIE::DmaCpyNdOp> maybeL3ToL2DmaOp =
+      fetchL3ToL2DmaCpyNdOp(l2ToL1DmaOps[0]);
+  if (failed(maybeL3ToL2DmaOp)) return failure();
+  l3ToL2DmaOps.push_back(maybeL3ToL2DmaOp.value());
+
+  // Check that all L3 buffer associated with the different L3->L2 Dma ops are
+  // same.
+  for (unsigned i = 1, n = l2ToL1DmaOps.size(); i < n; i++) {
+    maybeL3ToL2DmaOp = fetchL3ToL2DmaCpyNdOp(l2ToL1DmaOps[i]);
+    if (failed(maybeL3ToL2DmaOp)) return failure();
+    l3ToL2DmaOps.push_back(maybeL3ToL2DmaOp.value());
+    if (l3ToL2DmaOps[0].getSourceObjectFifo() !=
+        l3ToL2DmaOps[i].getSourceObjectFifo()) {
+      LLVM_DEBUG(llvm::dbgs()
+                 << "Found different L3 objectFifo for " << l3ToL2DmaOps[0]
+                 << " and " << l3ToL2DmaOps[i] << "\n");
+      return failure();
+    }
+    if (failed(checkIfSameDimensionalityAccessPatterns(l3ToL2DmaOps[0],
+                                                       l3ToL2DmaOps[i]))) {
+      LLVM_DEBUG(llvm::dbgs()
+                 << "Found different dimensionality of access patterns\n");
+      return failure();
+    }
+  }
+
+  if (l2ToL1DmaOps.size() != l3ToL2DmaOps.size()) {
+    LLVM_DEBUG(
+        llvm::dbgs()
+        << "expected 1:1 correspondence between L3->L2 and L2->L1 Dma ops\n");
+    return failure();
+  }
+
+  // Fetch split/non-split dimensions. Currently we look for a continuous
```

Comment:
Aren't these dimensions for combining instead of splitting? Also, it might be good to put the logic to find the splitting/combining dimensions in its own utility function.